File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def dfs(self,nodes,node,target):
        if not node:
            return False
        nodes.append(node)
        if node == target:
            return True
        if self.dfs(nodes,node.left,target) or self.dfs(nodes,node.right,target):
            return True
        nodes.pop()
        return False
        

    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        stack_p = []
        stack_q = []
        self.dfs(stack_p,root,p)
        self.dfs(stack_q,root,q)
        lca = None
        l = min(len(stack_p),len(stack_q))
     
        for i in range(l):
            if stack_p[i] == stack_q[i]:
                lca = stack_p[i]
            else:
                break
        return lca
        
        
        

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean createPath(TreeNode root,TreeNode p,ArrayList<TreeNode> result){
        if(root == null)
            return false;
        if(root == p){
            result.add(p);
            return true;
        }
        result.add(root);
        boolean left = createPath(root.left,p,result);
        boolean right = createPath(root.right,p ,result);
        if(!left && !right){
            result.remove(result.size() - 1);
            return false;
        }
        return true;




    }
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        ArrayList<TreeNode> path_p = new ArrayList<>();
        ArrayList<TreeNode> path_q = new ArrayList<>();
        int m;
        TreeNode result = null;
        createPath(root,p,path_p);

        createPath(root,q,path_q);

        m = Math.min(path_p.size(),path_q.size());
        for(int i = 0; i < m; i++){
            if(path_p.get(i) == path_q.get(i)){
                result = path_q.get(i);
            }
            else{
                break;
            }
        }
        return result;


 
        
    }
}